text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1401 @@
+
+2. In the absence of hard technical constraints forbidding it, we should **prefer the solution which has the best story for teaching**—at all levels, including beginners but also supporting advanced users. In particular, this means that we should value both *progressive disclosure of complexity* and the *principle of least surprise*, and that we may need to weight them against each other, but that we should pay particular attention when they agree.
+
+3. This design must cleanly interoperate with existing Ember codebases. That is, adopting this **must not require users to migrate their entire codebase at once**.
+
+4. We should **prefer a design which provides more flexibility** to end users over a design which provides less.
+
+While it is certainly possible to differ with these constraints *a priori*—reevaluating constraints is, in a very real sense, [how we got to this very RFC](https://github.com/emberjs/rfcs/pull/367#issuecomment-423839940)—we also run the risk of paralysis if we *continually* reevaluate from first principles. More challenging is inevitable disagreement about how we *weight* these constraints. On that front, there is no possibility of *final* agreement, but we should commit to some ordering for the purposes of this design so that the rest of it can proceed on the same terms.
+
+
+## Detailed design
+
+Introduce a new high-level syntax, the `<template>` tag, which is syntactical sugar for `setComponentTemplate` and `precompileTemplate`, in conjunction with the existing Ember and Glimmer `Component` classes and the special template-only component class returned by the `templateOnlyComponent` default export from `@ember/component/template-only`. There are three distinct, legal forms for this compilation:
+
+- a standalone `<template>` at the top level of a module
+- a `<template>` bound to a name
+- in the body of a component class
+
+For a discussion of the `setComponentTemplate` and `templateOnlyComponent` primitives, see [RFC #0481][rfc-0481]; for discussion of the `precompileTemplate` primitive, see [RFC #0496][rfc-0496]. This discussion will *assume* rather than *define* those.
+
+[rfc-0481]: https://emberjs.github.io/rfcs/0481-component-templates-co-location.html
+
+
+### Compilation
+
+The value produced by authoring a `<template>` is a *JavaScript value*, and accordingly may be exported, bound to other values, passed as an argument to a function or set as a value on a class, etc. However, the value may not be treated as *dynamic*: the rendering engine will evaluate the value exactly and only once. Therefore, it is nonsensical to use it with a `let` binding or to attempt to create new `<template>`s dynamically from the body of a function. (A function may of course return different components based on its arguments, etc.; it simply may not *construct* new component templates dynamically.)
```

Comment:
I'm still reading through, but wanted to comment here in case i forget.

Dynamic compilation is a necessity of REPLs, such as 
[https://limber.glimdown.com/][demo] (which uses the `<template>` syntax.

I don't think it's a relevant constraint that `<template>` cannot be dynamically created, because dynamic creation of templates is a gated possibility on the presence of the template compiler at runtime.

[demo]: https://limber.glimdown.com/?t=%23%20Limber%20Editor%0A%0A%23%23%20With%20JavaScript%0A%0Acode%20fence%20tags%20can%20be%20used%20to%20directly%20render%20components%2C%0Aas%20well%20as%20render%20the%20code%20snippet%20the%20live%20code%20comes%20from.%0A%0AOptions%3A%0A%0A%20-%20%60live%60%20-%20_renders%20in%20place_%0A%20-%20%60live%20preview%60%20_renders%20in%20place%2C%20placing%20the%20source%20after%20the%20live%20render_%0A%20-%20%60live%20preview%20below%60%20_renders%20in%20place%2C%20placing%20the%20live%20render%20below%20the%20source_%0A%0A%60%60%60gjs%20live%0A%3Ctemplate%3E%0A%20%20Hello%2C%20Glimmer!%0A%3C%2Ftemplate%3E%0A%60%60%60%0A%0A%60%60%60gjs%20live%20preview%0Aimport%20Component%20from%20%27%40glimmer%2Fcomponent%27%3B%0Aimport%20%7B%20tracked%20%7D%20from%20%27%40glimmer%2Ftracking%27%3B%0Aimport%20%7B%20on%20%7D%20from%20%27%40ember%2Fmodifier%27%3B%0A%0Aexport%20default%20class%20HelloWorld%20extends%20Component%20%7B%0A%20%20%40tracked%20count%20%3D%200%3B%0A%0A%20%20increment%20%3D%20()%20%3D%3E%20this.count%20%2B%3D%201%3B%0A%0A%20%20%3Ctemplate%3E%0A%20%20%20%20%3Cp%3EYou%20have%20clicked%20the%20button%20%7B%7Bthis.count%7D%7D%20times.%3C%2Fp%3E%0A%0A%20%20%20%20%3Cbutton%20%7B%7Bon%20%22click%22%20this.increment%7D%7D%3EClick%3C%2Fbutton%3E%0A%20%20%3C%2Ftemplate%3E%0A%7D%0A%60%60%60%0A

Reply:
Dynamic changes are possible—there’s a discussion of this in the context of using this proposal and the primitive s in function return values later—but the engine won’t reevaluate just because you changed a `let` binding somewhere. We would need to be compiling `let` into something else the way Svelte does for that to work!

Reply:
Maybe i got too hung up on the phrase 'construct dynamically'.

You're talking about the build time phase in thin paragraph, which is why nothing can be dynamic, and why you emphasize that 'let' example 🤔

Reply:
On rereading I see here what the confusion is. I’ll add a note (Monday) clarifying about needing the template compiler given that the normal mode of operation is to precompile during the build and remove the precompile invocation entirely from the runtime then. This more or less describes the authoring experience for normal app or addons devs and it *is* a constraint for that normal workflow and thus how we teach it, even if it isn’t a hard technical constraint.

Reply:
Updated with a more robust discussion of how scope is static for any given template compilation pass, and explained how it works with tools like REPLs/playgrounds/etc. which need to do new compilation in the app itself.